TESTS/mbed_drivers/sai/main.cpp

```diff
@@ -0,0 +1,293 @@
+static uint32_t receive_buffer[BUFFER_SIZE];
+
+/* Auxiliary function which sets the specified buffer using specified pattern. */
+static void set_buffer(void * buffer, uint32_t size, char pattern)
+{
+    char* p_byte = (char*) buffer;
+
+    while (size--) {
+        *p_byte = pattern;
+        p_byte++;
+    }
+}
+
+/* Auxiliary function which compares two buffers and returns true if both are the same. */
+static bool compare_buffers(const void *buffer1, const void *buffer2, uint32_t size)
+{
+    const unsigned char *p1 = (const unsigned char *) buffer1;
+    const unsigned char *p2 = (const unsigned char *) buffer2;
+    while (size--) {
+        if (*p1 != *p2) {
+            return false;
+        } else {
+            p1++;
+            p2++;
+        }
+    }
+
+    return true;
+}
+
+/* Number of formats which must be supported by SAI device.
+ * This variable defines number of obligate formats in test_sai_formats array. */
+const uint32_t required_formats_count = 5;
+
+/* Array of tested formats. */
+static sai_format_t test_sai_formats[] = {
+// Formats which are requested for all SAI devices
+//    sai_mode_i2s16,
+    sai_mode_i2s16w32,
+//    sai_mode_i2s32,
+//    sai_mode_pcm16l,
+//    sai_mode_pcm16s,
+// Other formats
+};
+
+/* Test that SAI class is able to communicate in loop-back mode.
+ *
+ * Given is SAI Class and SAI loop-back connection.
+ * When SAI object is configured as transmitter with the specified format and
+ *      another SAI object is configured as receiver with the same format used.
+ * Then transmitter can successfully transmit samples to the receiver.
+ */
+void SAI_transmission_test()
+{
+    loopback_test_tatus_t status;
+
+    if (!LOOPBACK_CONNECTION) {
+        TEST_IGNORE_MESSAGE("NO LOOPBACK CONNECTION - TEST SKIPPED!");
+        return;
+    }
+
+    char message[50];
+
+    for (uint32_t f_idx = 0; f_idx < (sizeof(test_sai_formats) / sizeof(sai_format_t)); f_idx++) {
+
+        SAI sai_transmitter(SAI_A_MCLK,                // mclk
+                SAI_A_BCLK,                // bclk
+                SAI_A_WCLK,                // wclk
+                SAI_A_SD,                  // SD
+                &test_sai_formats[f_idx],  // format
+                false,                     // is input
+                0,                         // master clock
+                false                      // internal mclk
+                );
+
+        SAI sai_receiver(SAI_B_MCLK,                 // mclk
+                SAI_B_BCLK,                 // bclk
+                SAI_B_WCLK,                 // wclk
+                SAI_B_SD,                   // SD
+                &test_sai_formats[f_idx],   // format
+                true,                       // is input
+                0,                          // master clock
+                false                       // internal mclk
+                );
+
+        // we should check init status before the actual test - for now there is no such option
```

Comment:
This is fine as is, just noting the following incoming PR to add the status check function:

https://github.com/ARMmbed/mbed-os/pull/6777